File: C++/Recursion/MaximumCuts.cpp
#include <bits/stdc++.h>
using namespace std;

int maxCuts(int n, int a, int b, int c)
{
  if (n < 0)
    return -1;
  if (n == 0)
    return 0;
  int res = 1 + max(maxCuts(n - a, a, b, c),
                    max(maxCuts(n - b, a, b, c), maxCuts(n - c, a, b, c)));
  return res;
}

int main()
{
  cout << maxCuts(5, 1, 2, 3) << endl;
}

File: Python/dynamic_programming/maximum_cuts.py
def max_cuts(n, a, b, c):
    dp = [-1] * (n + 1)
    dp[0] = 0
    for i in range(1, n + 1):
        if i - a >= 0:
            dp[i] = max(dp[i], dp[i - a])
        if i - b >= 0:
            dp[i] = max(dp[i], dp[i - b])
        if i - c >= 0:
            dp[i] = max(dp[i], dp[i - c])
        if dp[i] != -1:
            dp[i] += 1
    return dp[n]


if __name__ == "__main__":
    print(max_cuts(5, 1, 2, 3))
